src/sagemaker/jumpstart/session_utils.py

```diff
@@ -0,0 +1,162 @@
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""This module stores SageMaker Session utilities for JumpStart models."""
+
+from __future__ import absolute_import
+
+from typing import Optional, Tuple
+from sagemaker.jumpstart.constants import DEFAULT_JUMPSTART_SAGEMAKER_SESSION
+
+from sagemaker.jumpstart.utils import get_jumpstart_model_id_version_from_resource_arn
+from sagemaker.session import Session
+from sagemaker.utils import aws_partition
+
+
+def get_model_id_version_from_endpoint(
+    endpoint_name: str,
+    inference_component_name: Optional[str] = None,
+    sagemaker_session: Session = DEFAULT_JUMPSTART_SAGEMAKER_SESSION,
+) -> Tuple[str, str, Optional[str]]:
+    """Given an endpoint name (and optionally an IC name), get the JumpStart model ID and version.
+
+    Infers the model ID and version based on the resource tags. Returns a tuple of the model ID
+    and version. A third string element is included in the tuple for any inferred inference
+    component name, or None if it's a non-IC endpoint.
+
+    Raises:
+        ValueError: If model ID and version cannot be inferred from the endpoint.
+    """
+    if inference_component_name or sagemaker_session.is_ic_based_endpoint(endpoint_name):
+        if inference_component_name:
+            model_id, model_version = _get_model_id_version_from_ic_endpoint_with_ic_name(
+                inference_component_name, sagemaker_session
+            )
+
+        else:
+            (
+                model_id,
+                model_version,
+                inference_component_name,
+            ) = _get_model_id_version_from_ic_endpoint_without_ic_name(
+                endpoint_name, sagemaker_session
+            )
+
+    else:
+        model_id, model_version = _get_model_id_version_from_non_ic_endpoint(
+            endpoint_name, inference_component_name, sagemaker_session
+        )
+    return model_id, model_version, inference_component_name
+
+
+def _get_model_id_version_from_ic_endpoint_without_ic_name(
+    endpoint_name: str, sagemaker_session: Session
+) -> Tuple[str, str, str]:
+    """Given an endpoint name, derives the model ID, version, and inferred inference component name.
+
+    This function assumes the endpoint corresponds to an inference component-based endpoint.
+
+    Raises:
+        ValueError: If there is not a single inference component associated with the endpoint.
+    """
+    inference_components = sagemaker_session.list_inference_components(
+        endpoint_name_equals=endpoint_name
+    )["InferenceComponents"]
```

Comment:
question: do you need to paginate here? Is there a risk that the `list` utility returns an empty `InferenceComponents` list and a `NextToken` ?

Reply:
that's a good point, let me think about this case